### PR TITLE
Merge 'bitcoin/0.12', fix/dashify

### PR DESCRIPTION
**Pull request by UdjinM6**
Merge bitcoin upstream 0.12, dashify were needed, fix conflicts.

NOTE:
- .travis.yml is mostly preserved though dist changed to trusty, not sure if cache will still work,  probably needs additional attention and testing; @schinzelh pls have a closer look;
- release notes are not properly placed, need more work on this later + we should probably move our release notes from release-notes.md to another file because that one is "live";
- `mapAlreadyAskedFor` now uses inv.hash as a key instead of inv itself, could be a good idea to make sure that all hashes for our invs are actually unique.

### CODE REVIEW

**Comment by tgflynn** [comment]
What's the relationship  between this and this previous commit: https://github.com/dashpay/dash/pull/914 ?

Are we trying to sync with 2 different upstreams ?

**Comment by UdjinM6** [comment]
@tgflynn Bitpay's bitcoin is a fork of original bitcoin 0.12, it has additional feature on top of original and is optional. This merges original upstream which is more or less a must.

**Comment by tgflynn** [comment]
So Bitpay's version doesn't itself track core ?

**Comment by UdjinM6** [comment]
@tgflynn like for any other fork it's completely up to them if they want to track it or not, we can't force them to do this the way we want :) Bitpay's version was more like "a one time merge of cool additional features" and I would keep it this way ~~merging~~ cherry-picking only critical patches for their features while merging the whole original upstream regardless of Bitpay's vision of tracking it.

**Comment by tgflynn** [comment]
I take it there will still be another set of changes to merge from 0.12.1 ?

**Comment by tgflynn** [comment]
What about this commit ?

https://github.com/dashpay/dash/commit/025f0901c43f99166313f9bbc53332cee5c98962

"This brings Dash 0.12.1.x codebase in line with Bitcoin release 0.12.1" - 56 changed files.

**Comment by UdjinM6** [comment]
@tgflynn There is no 0.12.1 branch in bitcoin, they push everything 0.12.x to 0.12 (note 0.12.2. in few commits here already). 025f090 i.e. #897 is 1.5 month old, this bring new commits and aligns our codebase again.

Pushed 790607b to fix tests.

**Comment by tgflynn** [comment]
Then these changes must be targeting bitcoin 0.13 because 0.12.1 was the last 0.12 release.

**Comment by UdjinM6** [comment]
https://github.com/bitcoin/bitcoin/tree/0.12

**Comment by tgflynn** [comment]
The last commit there was from July 11.  I imagine they decided not to release 0.12.2 sometime between then and the 0.13 release.  I certainly don't think they're going to do it now that 0.13 has been released.  So I think these are changes that will either never go into a bitcoin release or will/have gone into 0.13 for the ones that were back-ported.

File: src/core_write.cpp [hunk 1]
```diff
@@ -45,7 +45,7 @@ string FormatScript(const CScript& script)
             if (vch.size() > 0) {
                 ret += strprintf("0x%x 0x%x ", HexStr(it2, it - vch.size()), HexStr(it - vch.size(), it));
             } else {
-                ret += strprintf("0x%x", HexStr(it2, it));
+                ret += strprintf("0x%x ", HexStr(it2, it));
```
Comment (tgflynn) [inline src/core_write.cpp]: This looks strange.  HexStr returns a string but %x format normally expects an int type.

File: src/main.cpp [hunk 2]
```diff
@@ -5308,11 +5308,11 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                 CAddress addr = GetLocalAddress(&pfrom->addr);
                 if (addr.IsRoutable())
                 {
-                    LogPrintf("ProcessMessages: advertizing address %s\n", addr.ToString());
+                    LogPrintf("ProcessMessages: advertising address %s\n", addr.ToString());
```
Comment (tgflynn) [inline src/main.cpp]: :+1: Those were starting to get annoying.

**Comment by evan82** [comment]
Thanks!

**Pull request merged**